File: snippets/go/usecases/microservices/orderservice.go
package main

import (
	restate "github.com/restatedev/sdk-go"
)

type Order struct {
	ID         string `json:"id"`
	Items      []Item `json:"items"`
	CreditCard string `json:"credit_card"`
}

type Item struct {
	ID       string `json:"id"`
	Quantity int    `json:"quantity"`
}

type OrderResult struct {
	Success   bool   `json:"success"`
	PaymentID string `json:"payment_id"`
}

func ChargePayment(creditCard, paymentID string) (restate.Void, error) {
	// Simulate payment processing
	return restate.Void{}, nil
}

func ReserveInventory(itemID string, quantity int) (restate.Void, error) {
	// Simulate payment processing
	return restate.Void{}, nil
}

// <start_here>
type OrderService struct{}

func (OrderService) Process(ctx restate.Context, order Order) (OrderResult, error) {
	// Each step is automatically durable and resumable
	paymentID := restate.Rand(ctx).UUID().String()

	_, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return ChargePayment(order.CreditCard, paymentID)
	})
	if err != nil {
		return OrderResult{}, err
	}

	for _, item := range order.Items {
		_, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
			return ReserveInventory(item.ID, item.Quantity)
		})
		if err != nil {
			return OrderResult{}, err
		}
	}

	return OrderResult{Success: true, PaymentID: paymentID}, nil
}

// <end_here>

File: snippets/python/src/usecases/microservices/order_service.py
import restate
from typing import List, Dict, Any
import uuid

from pydantic import BaseModel


class Item(BaseModel):
    id: str
    amount: int


class Order(BaseModel):
    id: str
    items: List[Item]
    credit_card: str


def charge_payment(credit_card: str, payment_id: str):
    # Simulate payment processing
    pass


def reserve(item_id: str, amount: int):
    # Simulate inventory reservation
    pass


# <start_here>
order_service = restate.Service("OrderService")


@order_service.handler()
async def process(ctx: restate.Context, order: Order):
    # Each step is automatically durable and resumable
    payment_id = str(uuid.uuid4())

    await ctx.run_typed(
        "charge", charge_payment, credit_card=order.credit_card, payment_id=payment_id
    )

    for item in order.items:
        await ctx.run_typed(
            f"reserve_{item.id}", reserve, item_id=item.id, amount=item.amount
        )

    return {"success": True, "payment_id": payment_id}


# <end_here>
